# 実施日時　2023/10/23-2023/10/24
## 動作環境
- python 3.11.5
- psycopg2==2.9.9
- alembic==1.12.0
- SQLAlchemy==2.0.22
- sqlalchemy-seeder==0.3.0
## 概要
SQLiteからpostgresへの移行の手順をログとして残す

## .envファイルの作成
.env.sampleに基づいて.envのテンプレートを作成する
```
PROJECT_NAME="FastAPI+SQLArchemy Todo Sample"
API_V1_STR="/v1"
BACKEND_CORS_ORIGINS="localhost"
DATABASE_URL="sqlite:///todo.db"
LOGGING_CONF="./logging.json"
```

`postgresql://<username>:<password>@<hostname>:<port>/<dbname>`
上記の形式に沿ってDATABASE_URLを書き換える


## テーブル作成用のスクリプトを作成する(上手く行かなかった)

作業
```
#スクリプト作成
$ alembic revision -m "Creat Table"
bash: alembic: command not found

#alembicが入っていなかったため、インストール
$ pip install alembic
Installing collected packages: typing-extensions, MarkupSafe, greenlet, SQLAlchemy, Mako, alembic
Successfully installed Mako-1.2.4 MarkupSafe-2.1.3 SQLAlchemy-2.0.22 alembic-1.12.0 greenlet-3.0.0 typing-extensions-4.8.0

#alembic.iniが既存のため、alembic initはしない
$ alembic revision -m "Creat Table"
Generating /home/codeserver/todo-team-b/practical-fastapi/migration/versions/21c8012d2c97_creat_table.py ...  done
```

## 作成するテーブル情報を定義する
先ほどのコマンドで生成された21c8012d2c97_creat_table.pyに対してテーブルの情報を書き込む
```
"""Creat Table

Revision ID: 21c8012d2c97
Revises: 10d0eb272386
Create Date: 2023-10-23 15:24:31.198952

"""
from alembic import op
import sqlalchemy as sa
from sqlalchemy.dialects import postgresql as pg


# revision identifiers, used by Alembic.
revision = '21c8012d2c97'
down_revision = '10d0eb272386'
branch_labels = None
depends_on = None


def upgrade() -> None:
    # ### commands auto generated by Alembic - please adjust! ###
    op.create_table('tag',
    sa.Column('id', pg.INTEGER, nullable=False),
    sa.Column('name', pg.VARCHAR(32), nullable=False),
    sa.Column('created_at', pg.TIMESTAMP(timezone=True), server_default=sa.text("(DATETIME('now', 'localtime'))"), nullable=False),
    sa.Column('updated_at', pg.TIMESTAMP(timezone=True), server_default=sa.text("(DATETIME('now', 'localtime'))"), nullable=False),
    sa.PrimaryKeyConstraint('id'),
    sa.UniqueConstraint('name')
    )
    op.create_index(op.f('ix_tag_id'), 'tag', ['id'], unique=False)
    op.create_table('todo',
    sa.Column('id', pg.INTEGER, nullable=False),
    sa.Column('content', pg.VARCHAR(256), nullable=False),
    sa.Column('completed', pg.BOOLEAN, server_default='False', nullable=False),
    sa.Column('deadline', pg.TIMESTAMP, nullable=True),
    sa.Column('created_at', pg.TIMESTAMP(timezone=True), server_default=sa.text("(DATETIME('now', 'localtime'))"), nullable=False),
    sa.Column('updated_at', pg.TIMESTAMP(timezone=True), server_default=sa.text("(DATETIME('now', 'localtime'))"), nullable=False),
    sa.PrimaryKeyConstraint('id')
    )
    op.create_index(op.f('ix_todo_id'), 'todo', ['id'], unique=False)
    op.create_table('todo_tag',
    sa.Column('todo_id', pg.INTEGER, nullable=False),
    sa.Column('tag_id', pg.INTEGER, nullable=False),
    sa.Column('created_at', pg.TIMESTAMP(timezone=True), server_default=sa.text("(DATETIME('now', 'localtime'))"), nullable=False),
    sa.Column('updated_at', pg.TIMESTAMP(timezone=True), server_default=sa.text("(DATETIME('now', 'localtime'))"), nullable=False),
    sa.ForeignKeyConstraint(['tag_id'], ['tag.id'], ),
    sa.ForeignKeyConstraint(['todo_id'], ['todo.id'], ),
    sa.PrimaryKeyConstraint('todo_id', 'tag_id')
    )
    # ### end Alembic commands ###


def downgrade() -> None:
    # ### commands auto generated by Alembic - please adjust! ###
    op.drop_table('todo_tag')
    op.drop_index(op.f('ix_todo_id'), table_name='todo')
    op.drop_table('todo')
    op.drop_index(op.f('ix_tag_id'), table_name='tag')
    op.drop_table('tag')

```

## テーブルを作成する
```
$ alembic upgrade head
ModuleNotFoundError: No module named 'psycopg2'
```

↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑↑できなかった↑↑↑↑↑↑↑↑↑↑↑↑↑

## 手段２　modelsの方を直してautogenerateの方針

### 完了したので操作手順をまとめる

まず、SQLiteからpostgresに変更したら発生した問題
```bash
sqlalchemy.exc.ProgrammingError: (psycopg2.errors.UndefinedFunction) 関数datetime(unknown, unknown)は存在しません
LINE 7:  created_at TIMESTAMP WITH TIME ZONE DEFAULT DATETIME('now',...
                                                     ^
HINT:  指定した名前と引数型に合致する関数がありません。明示的な型変換が必要かもしれません。

[SQL: 
CREATE TABLE todo (
        id SERIAL NOT NULL, 
        content VARCHAR(256) NOT NULL, 
        completed BOOLEAN DEFAULT 'False' NOT NULL, 
        deadline TIMESTAMP WITHOUT TIME ZONE, 
        created_at TIMESTAMP WITH TIME ZONE DEFAULT DATETIME('now', 'localtime') NOT NULL, 
        updated_at TIMESTAMP WITH TIME ZONE DEFAULT DATETIME('now', 'localtime') NOT NULL, 
        PRIMARY KEY (id)
)

]
(Background on this error at: https://sqlalche.me/e/20/f405)
```

SQLiteとpostgresでのDATE型の扱いが違ったために起きた
解決方法：[Zenn:alembicのマイグレーション自動生成をカスタマイズ](https://zenn.dev/sendo_tech/articles/b859e1fe816946)
上記を参考にすると日時は以下のように指定している
```python
class TimestampMixin(object):
    create_date: Mapped[datetime] = mapped_column(postgresql.TIMESTAMP(timezone=True),
                                                  default=lambda: datetime.now(ZoneInfo("Asia/Tokyo")),
                                                  comment='作成日')
    update_date: Mapped[datetime] = mapped_column(postgresql.TIMESTAMP(timezone=True),
                                                  nullable=True,
                                                  onupdate=lambda: datetime.now(ZoneInfo("Asia/Tokyo")),
                                                  comment='更新日')
```

この書き方を参考にmodels配下、tag.py,todo.py,todo_tag.pyを書き直す
書き直したものは以下のようになる。
```python
from datetime import datetime
from zoneinfo import ZoneInfo

from sqlalchemy.dialects import postgresql
~~~~~~~~~~~~~~~~~~~~~~~~~~~
~~~~~~~~~~~~~~~~~~~~~~~~~~~
    created_at: Mapped[datetime] = mapped_column(
        postgresql.TIMESTAMP(timezone=True),
        default=lambda: datetime.now(ZoneInfo("Asia/Tokyo")),
    )

    updated_at: Mapped[datetime] = mapped_column(
        postgresql.TIMESTAMP(timezone=True),
        onupdate=lambda: datetime.now(ZoneInfo("Asia/Tokyo")),
    )
```

ここまで出来たが、alembicで反映させようとするとエラーが出た。
```bash
$ alembic revision --autogenera
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
ERROR [alembic.util.messaging] Target database is not up to date.
  FAILED: Target database is not up to date.
```

おそらく読み込んでいるものが、一個手前のバージョンになっているため、エラーが出た。
Script文の余分なコードを消して初期状態に戻す。
```
"""Creat Table

Revision ID: 21c8012d2c97
Revises: 10d0eb272386
Create Date: 2023-10-23 15:24:31.198952

"""
from alembic import op
import sqlalchemy as sa
from sqlalchemy.dialects import postgresql as pg


# revision identifiers, used by Alembic.
revision = '21c8012d2c97'
down_revision = '10d0eb272386'
branch_labels = None
depends_on = None


def upgrade() -> None:
    # ### commands auto generated by Alembic - please adjust! ###
    pass
    # ### end Alembic commands ###


def downgrade() -> None:
    # ### commands auto generated by Alembic - please adjust! ###
    pass
```

戻した後に、もう一度操作をすると上手く動作した。
```
codeserver@webapp2:~/todo-team-b/practical-fastapi$ alembic upgrade head
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade 21c8012d2c97 -> 3007845c0896, empty message
codeserver@webapp2:~/todo-team-b/practical-fastapi$ alembic revision --autogenerate
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.autogenerate.compare] Detected added table 'tag'
INFO  [alembic.autogenerate.compare] Detected added index 'ix_tag_id' on '['id']'
INFO  [alembic.autogenerate.compare] Detected added table 'todo'
INFO  [alembic.autogenerate.compare] Detected added index 'ix_todo_id' on '['id']'
INFO  [alembic.autogenerate.compare] Detected added table 'todo_tag'
  Generating /home/codeserver/todo-team-b/practical-fastapi/migration/versions/947de0195030_.py ...  done
codeserver@webapp2:~/todo-team-b/practical-fastapi$ alembic upgrade head
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade 3007845c0896 -> 947de0195030, empty message
```

拡張機能から確認したら無事にテーブルが作成されていた。